### ISSUE TITLE
Frequent disconnection with EXPIRED_SAS_TOKEN

### ISSUE THREAD

**Issue by rtheil-growlink** [Azure/azure-iot-sdk-c#2274]
Seeing an issue that has popped up lately that I'm not sure I can fix, but haven't seen before.

Client is connected ok, will even get a callback for success received message, then will suddenly have an expired SAS token and reconnect. It always reconnects successfully but this process seems to happen quite often. I see 153 since Friday on one of my devices. Also am noticing that it seems to happen on a consistent schedule of every 48 minutes on multiple pieces of hardware. A very strange interval.

Thanks for all the assistance with this stuff! You guys rock!

```
I (2879396) AzureService: Confirmation[149] received for message tracking id = 149 with result = IOTHUB_CLIENT_CONFIRMATION_OK

-> 07:35:14 DISCONNECT

W (2889046) AzureService: 

Connection Status result:IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, Connection Status reason: IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN

I (2894856) gpio: GPIO[12]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 [0m
-> 07:35:25 CONNECT | VER: 4 | KEEPALIVE: 60 | FLAGS: 192 | USERNAME: xxxxxxxxxxx.azure-devices.net/0000000000000b30/?api-version=2020-09-30&DeviceClientType=iothubclient%2f1.8.0%20(native%3b%20freertos%3b%20esp%20platform) | PWD: XXXX | CLEAN: 0

W (2904886) AzureService: 

Connection Status result:IOTHUB_CLIENT_CONNECTION_AUTHENTICATED, Connection Status reason: IOTHUB_CLIENT_CONNECTION_OK

```

**Comment by rtheil-growlink** [comment]
Thanks, Dane. So really, I need to handle these cases where the token expires differently. Right now I'm assuming anything that isn't IOTHUB_CLIENT_CONNECTION_OK is a disconnection event, and updating logs and a hardware LED accordingly. But it sounds like this isn't the kind of disconnection event I need to worry about since it happens about every hour.

What's the behavior if SAS token expires and subsequently can't be refreshed? Will there be another call to the ConnectionStatusCallback method?

**Comment by danewalton** [comment]
The SAS token will be re-upped as you call the DoWork function. The log on the ConnectionStatusCallback helps you to understand the state of the device at the time. But if the reason is "expired SAS token", the SDK will make an attempt to get a new one. If the token can't be refreshed, you should get another status update from the callback with further details as to why. Those scenarios might need to be handled differently.

TLDR is it should recover from that disconnection.

**Comment by rtheil-growlink** [comment]
Thanks for all that info. Will take a much closer look at that whole process as I believe I have it oversimplified at the moment. Closing.

**Comment by rtheil-growlink** [comment]
Hey, Dane. Is there any way to force the SAS token to expire and require the client to reup on the next doWork call? I've found an issue related to how Espressif wrote their port for this SDK, so when the device loses connection to the wifi AP, TLS has immediate issues when reconnected wo wifi and can't recover. However, I found that when the SAS token expires the Azure client is able to get the connection re-established. I'd like to force-expire the SAS token the moment there's a loss of AP connectivity detected in order to get past this issue.
